// dom-ease/dist/ease.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var EventEmitter = require('eventemitter3');

var Number = require('./number');
var Color = require('./color');
var Transform = require('./transform');
var Margin = require('./margin');
var utils = require('./utils');

var Ease = function (_EventEmitter) {
    _inherits(Ease, _EventEmitter);

    /**
     * Ease class returned by DomEase.add()
     * @extends EventEmitter
     * @param {HTMLElement} element
     * @param {object} params
     * @param {number} [params.left] in px
     * @param {number} [params.right] in px
     * @param {number} [params.top] in px
     * @param {number} [params.bottom] in px
     * @param {number} [params.width] in px
     * @param {number} [params.height] in px
     * @param {number} [params.scale]
     * @param {number} [params.scaleX]
     * @param {number} [params.scaleY]
     * @param {number} [params.opacity]
     * @param {number} [params.marginLeft] in px
     * @param {number} [params.marginRight] in px
     * @param {number} [params.marginTop] in px
     * @param {number} [params.marginBottom] in px
     * @param {(color|color[])} [params.color]
     * @param {(color|color[])} [params.backgroundColor]
     * @param {object} [options]
     * @param {number} [options.start] use this as the starting value
     * @param {number} [options.duration]
     * @param {(string|function)} [options.ease]
     * @param {(boolean|number)} [options.repeat]
     * @param {boolean} [options.reverse]
     * @param {number} [options.wait]
     * @returns {Ease}
     * @fires Ease#each
     * @fires Ease#complete
     * @fires Ease#loop
     * @hideconstructor
     */
    function Ease(element, params, options) {
        _classCallCheck(this, Ease);

        var _this = _possibleConstructorReturn(this, (Ease.__proto__ || Object.getPrototypeOf(Ease)).call(this));

        _this.element = element;
        _this.list = [];
        _this.time = 0;
        _this.duration = options.duration;
        _this.ease = options.ease;
        _this.repeat = options.repeat;
        _this.reverse = options.reverse;
        _this.wait = options.wait || 0;
        for (var entry in params) {
            switch (entry) {
                case 'left':
                    _this.list.push(new Number(element, entry, element.offsetLeft, params[entry], 'px'));
                    break;

                case 'top':
                    _this.list.push(new Number(element, entry, element.offsetTop, params[entry], 'px'));
                    break;

                case 'bottom':
                    _this.list.push(new Number(element, entry, element.parentNode.offsetHeight - (element.offsetTop + element.offsetHeight), params[entry], 'px'));
                    break;

                case 'right':
                    _this.list.push(new Number(element, entry, element.parentNode.offsetWidth - (element.offsetLeft + element.offsetWidth), params[entry], 'px'));
                    break;

                case 'color':
                    _this.list.push(new Color(element, 'color', params[entry], _this.duration / (1 + params[entry].length)));
                    break;

                case 'backgroundColor':
                    _this.list.push(new Color(element, 'backgroundColor', _this.duration / (1 + params[entry].length)));
                    break;

                case 'scale':
                    if (_this.transform) {
                        _this.transform.add('scaleX', params[entry]);
                    } else {
                        _this.transform = new Transform(element, 'scaleX', params[entry]);
                        _this.list.push(_this.transform);
                    }
                    _this.transform.add('scaleY', params[entry]);
                    break;

                case 'scaleX':
                case 'scaleY':
                    if (_this.transform) {
                        _this.transform.add(entry, params[entry]);
                    } else {
                        _this.transform = new Transform(element, entry, params[entry]);
                        _this.list.push(_this.transform);
                    }
                    break;

                case 'opacity':
                    _this.list.push(new Number(element, entry, parseFloat(utils.getComputed(element, 'opacity')), params[entry]));
                    break;

                case 'width':
                    _this.list.push(new Number(element, entry, element.offsetWidth, params[entry], 'px'));
                    break;

                case 'height':
                    _this.list.push(new Number(element, entry, element.offsetHeight, params[entry], 'px'));
                    break;

                case 'marginLeft':
                case 'marginRight':
                case 'marginTop':
                case 'marginBottom':
                    if (_this.margin) {
                        _this.margin.add(entry, params[entry]);
                    } else {
                        _this.margin = new Margin(element, entry, params[entry]);
                        _this.list.push(_this.margin);
                    }
                    break;

                default:
                    console.warn(entry + ' not setup for animation in dom-ease.');
            }
        }
        return _this;
    }

    _createClass(Ease, [{
        key: 'update',
        value: function update(elapsed) {
            if (this.wait) {
                this.wait -= elapsed;
                if (this.wait < 0) {
                    elapsed = -this.wait;
                    this.wait = 0;
                } else {
                    return;
                }
            }
            var list = this.list;
            var leftover = null;
            this.time += elapsed;
            if (this.time >= this.duration) {
                leftover = this.time - this.duration;
                this.time -= leftover;
            }
            var percent = this.ease(this.time, 0, 1, this.duration);
            for (var i = 0, _i = list.length; i < _i; i++) {
                list[i].update(percent, this.time);
            }
            this.emit('each', this);

            // handle end of duration
            if (leftover !== null) {
                if (this.reverse) {
                    this.reverseEases();
                    this.time = leftover;
                    this.emit('loop', this);
                    if (!this.repeat) {
                        this.reverse = false;
                    } else if (this.repeat !== true) {
                        this.repeat--;
                    }
                } else if (this.repeat) {
                    this.emit('loop', this);
                    this.time = leftover;
                    if (this.repeat !== true) {
                        this.repeat--;
                    }
                } else {
                    this.emit('complete', this);
                    return true;
                }
            }
        }
    }, {
        key: 'reverseEases',
        value: function reverseEases() {
            var list = this.list;
            for (var i = 0, _i = list.length; i < _i; i++) {
                var ease = list[i];
                ease.reverse();
            }
        }
    }]);

    return Ease;
}(EventEmitter);

/**
 * fires when eases are complete
 * @event Ease#complete
 * @type {Ease}
 */

/**
 * fires on each loop while eases are running
 * @event Ease#each
 * @type {Ease}
 */

/**
 * fires when eases repeat or reverse
 * @event Ease#loop
 * @type {Ease}
 */

module.exports = Ease;
